crates/matrix-sdk/src/http_client.rs

```diff
@@ -455,9 +493,23 @@ impl HttpSend for reqwest::Client {
         &self,
         request: http::Request<Bytes>,
         _timeout: Duration,
+        send_progress: SharedObservable<TransmissionProgress>,
     ) -> Result<http::Response<Bytes>, HttpError> {
         #[allow(unused_mut)]
-        let mut request = reqwest::Request::try_from(request)?;
+        let mut request = if send_progress.subscriber_count() != 0 {
+            send_progress.update(|p| p.total += request.body().len());
+            reqwest::Request::try_from(request.map(|body| {
+                let chunks = stream::iter(BytesChunks::new(body, 8192).map(
+                    move |chunk| -> Result<_, Infallible> {
+                        send_progress.update(|p| p.current += chunk.len());
+                        Ok(chunk)
+                    },
+                ));
+                reqwest::Body::wrap_stream(chunks)
+            }))?
+        } else {
+            reqwest::Request::try_from(request)?
+        };
```

Comment:
@poljar this is what's required for tracking upload progress with reqwest. I think anybody implementing their own `HttpSend` would have a really hard time figuring out the equivalent solution for whatever library they are using, and default to not tracking progress at all. Do you think we could drop support for different HTTP transports, and instead allow users to configure the `reqwest::Client` instance used by the SDK client?

Reply:
It sounds a bit scary to remove it, but I see your point, things becomes quite library specific.

As far as I know, the only requirement we have is that people should be able to customize the HTTP headers each request sends out, so we should be able to do this. I guess if we do end up needing it we can come back and think of a better abstraction for this.